### PR TITLE
HIVE-26203: Implement alter iceberg table metadata location

### PR DESCRIPTION
**Pull request by lcspinter**
### What changes were proposed in this pull request?
Allow altering the metadata location of an iceberg table. If the new metadata belongs to another table raise an exception.

### Does this PR introduce _any_ user-facing change?
The user can execute 
`ALTER TABLE tbl_name SET TBLPROPERTIES('metadata_location'='')`

### How was this patch tested?
Unit test

### CODE REVIEW

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java [hunk 1]
```diff
@@ -336,6 +336,30 @@ public void preAlterTable(org.apache.hadoop.hive.metastore.api.Table hmsTable, E
       // that users can change data types or reorder columns too with this alter op type, so its name is misleading..)
       assertNotMigratedTable(hmsTable.getParameters(), "CHANGE COLUMN");
       handleChangeColumn(hmsTable);
+    } else if (AlterTableType.ADDPROPS.equals(currentAlterTableOp)) {
+      assertNotThirdPartyMetadataLocationChange(hmsTable.getParameters());
```
Comment (pvary) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:340]: nit: `crossTableMetadataLocationChange`?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java [hunk 2]
```diff
@@ -336,6 +336,30 @@ public void preAlterTable(org.apache.hadoop.hive.metastore.api.Table hmsTable, E
...
+  }
+
+  /**
+   * Perform a check on the current iceberg table whether a metadata change can be performed. A table is eligible if
+   * the current metadata uuid and the new metadata uuid matches.
+   * @param tblParams hms table properties, must be non-null
+   */
+  private void assertNotThirdPartyMetadataLocationChange(Map<String, String> tblParams) {
+    if (tblParams.containsKey(BaseMetastoreTableOperations.METADATA_LOCATION_PROP)) {
+      Preconditions.checkArgument(icebergTable != null,
```
Comment (pvary) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:351]: How could this happen?

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerNoScan.java [hunk 3]
```diff
@@ -1456,6 +1456,63 @@ public void testCreateTableWithMetadataLocation() throws IOException {
             HiveIcebergStorageHandlerTestUtils.CUSTOMER_RECORDS.stream()).collect(Collectors.toList()), records, 0);
   }
 
+  @Test
+  public void testAlterTableWithMetadataLocation() throws IOException {
+    Assume.assumeTrue("Alter table with metadata location is only supported for Hive Catalog tables",
```
Comment (pvary) [inline iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerNoScan.java:1461]: What happens in the code where we try to do this for a non-HiveCatalog table?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java (hunk 2 above)
Comment (pvary) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:351]: Can we reword the error message to help the user to understand the situation better?
Like:
```
Can not perform table migration to Iceberg and setting the snapshot location in one step. Please migrate the table first
```

Or something like this

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerNoScan.java (hunk 3 above)
Comment (lcspinter) [reply iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerNoScan.java:1461]: They will see no errors, the property will be changed on the HMS table, but the current metadata won't change.

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java (hunk 2 above)
Comment (lcspinter) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:351]: Sure, that can be done. I will add a unit test as well.

**Comment by RoyalFeng** [comment]
@lcspinter  @pvary  @kgyrtkirk  @toddlipcon  @massdosage  can you help me see this problem, Thank You

https://stackoverflow.com/questions/72186228/hive-fetch-result-from-hdfs-is-too-slow-because-of-too-many-the-map-only-task

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerNoScan.java (hunk 3 above)
Comment (pvary) [reply iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerNoScan.java:1461]: Seems like an inconsistent state.
Is there a way to recover from that, or shall we throw an exception instead?

Comment (lcspinter) [reply iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerNoScan.java:1461]: I'm not sure this is inconsistent. The HMS table property `metadata_location` has a special meaning if the table belongs to the hive catalog. It is also treated differently in our `HiveTableOperations` class, refreshing the metadata based on the value of this property.
In the case of tables from other catalog types, this is just a regular table property and it doesn't have any highlighted meaning. So I run an alter query I would expect that the table property is changed but nothing else.

**Review by pvary — APPROVED** [review]

**Pull request merged**